source/opt/scalar_replacement_pass.cpp

```diff
@@ -0,0 +1,598 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "scalar_replacement_pass.h"
+
+#include "make_unique.h"
+#include "reflect.h"
+
+#include <queue>
+
+namespace spvtools {
+namespace opt {
+
+// Heuristic aggregate element limit.
+const uint32_t MAX_NUM_ELEMENTS = 100u;
+
+Pass::Status ScalarReplacementPass::Process(ir::IRContext* c) {
+  InitializeProcessing(c);
+
+  Status status = Status::SuccessWithoutChange;
+  for (auto& f : *get_module()) {
+    Status functionStatus = ProcessFunction(&f);
+    if (functionStatus == Status::Failure)
+      return functionStatus;
+    else if (functionStatus == Status::SuccessWithChange)
+      status = functionStatus;
+  }
+
+  return status;
+}
+
+Pass::Status ScalarReplacementPass::ProcessFunction(ir::Function* function) {
+  std::queue<ir::Instruction*> worklist;
+  ir::BasicBlock& entry = *function->begin();
+  for (auto iter = entry.begin(); iter != entry.end(); ++iter) {
+    // Function storage class OpVariables must appear as the first instructions
+    // of the entry block.
+    if (iter->opcode() != SpvOpVariable) break;
+
+    ir::Instruction* varInst = &*iter;
+    if (CanReplaceVariable(varInst)) {
+      worklist.push(varInst);
+    }
+  }
+
+  Status status = Status::SuccessWithoutChange;
+  while (!worklist.empty()) {
+    ir::Instruction* varInst = worklist.front();
+    worklist.pop();
+
+    if (!ReplaceVariable(varInst, &worklist))
+      return Status::Failure;
+    else
+      status = Status::SuccessWithChange;
+  }
+
+  return status;
+}
+
+bool ScalarReplacementPass::ReplaceVariable(
+    ir::Instruction* inst, std::queue<ir::Instruction*>* worklist) {
+  std::vector<ir::Instruction*> replacements;
+  CreateReplacementVariables(inst, &replacements);
+
+  bool ok = true;
+  std::vector<ir::Instruction*> dead;
+  dead.push_back(inst);
+  get_def_use_mgr()->ForEachUser(
+      inst, [this, &ok, replacements, &dead](ir::Instruction* user) {
```

Comment:
Looks like a value capture of |replacements|, could this cause many call to the copy constructor for the vector?

Reply:
Good catch.